### PR TITLE
fix: 24435 Fixed `cache-cleaner` thread leak

### LINKED ISSUE #24435: Virtual node cache cleaning threads pile up over time
This seems to be a regression from the following recent change: https://github.com/hiero-ledger/hiero-consensus-node/pull/24061/

The symptoms are: the number of `<virtual-map: cache-cleaner #?>` threads in a process keeps increasing over time. This may happen, because the cleaning pool is created right in VirtualNodeCache constructor, the core pool size is not 0, and some node cache instances are never destroyed with the `shutdown()` calls (e.g. node cache snapshots).

### PR DESCRIPTION
**Pull request by imalygin**
### Problem

Moving `cleaningPool` from static to instance level introduced a thread leak. `snapshot()` calls the primary constructor, creating a new `ThreadPoolExecutor` each time. These pools are never shut down in:

- **`VirtualMap.createSnapshot()`** — snapshot cache is flushed and dropped, pool never shut down
- **`VirtualMap.setupWithOriginalNode()`** — same pattern: snapshot, flush, drop
- **`VirtualMap.detach()`** — snapshot wrapped in `RecordAccessor`, whose `close()` only closes the data source

Core threads persist indefinitely (`allowCoreThreadTimeOut` defaults to `false`), so every call leaks one pool.

### Fix

`snapshot()` → `snapshot(boolean needsCleaningPool)`:

- **`snapshot(true)`** — inherits the parent's `cleaningPool`. Used by `createSnapshot()` and `setupWithOriginalNode()`, which call `flush()` → `deletedLeaves()` → `parallelTraverse(cleaningPool, ...)`.
- **`snapshot(false)`** — sets `cleaningPool` to `null`. Used by `detach()` (lookup-only). Methods requiring the pool (`deletedLeaves()`, `purge()`) assert non-null, so misuse fails fast.

Additional safety:

1. **Explicit `shutdown()` calls** — `createSnapshot()` finally block, `setupWithOriginalNode()` after flush, `RecordAccessor.close()`, `VirtualNodeCacheTest @AfterEach`.
2. **`allowCoreThreadTimeOut(true)`** on the primary constructor's pool — idle threads reclaimed after 60s even if `shutdown()` is missed.

**Related issue(s)**:

Fixes #24435 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 1]
```diff
@@ -311,21 +311,25 @@ public VirtualNodeCache(
...
-                                .setComponent("virtual-map")
-                                .setThreadName("cache-cleaner")
-                                .setExceptionHandler((t, ex) -> logger.error(
-                                        EXCEPTION.getMarker(), "Failed to purge unneeded key/mutationList pairs", ex))
-                                .buildFactory());
+        if (Boolean.getBoolean("syncCleaningPool")) {
+            cleaningPool = Runnable::run;
+        } else {
+            final ThreadPoolExecutor pool = new ThreadPoolExecutor(
+                    virtualMapConfig.getNumCleanerThreads(),
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:318]: Do you think setting the core pool size to 0 would be helpful? The number of tasks in this pool is relatively low, so overhead from creating new threads when needed should be negligible.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 2]
```diff
@@ -939,8 +972,8 @@ public long getFastCopyVersion() {
      */
     public VirtualNodeCache snapshot() {
         synchronized (lastReleased) {
-            final VirtualNodeCache newSnapshot =
-                    new VirtualNodeCache(virtualMapConfig, hashChunkHeight, hashChunkLoader, fastCopyVersion.get());
+            final VirtualNodeCache newSnapshot = new VirtualNodeCache(
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:965]: This snapshot is not expected to run any jobs in the cleaning pool. It should be possible to mark `cleaningPool` as nullable and just set it to null here

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 3]
```diff
@@ -311,21 +311,25 @@ public VirtualNodeCache(
...
-                        new LinkedBlockingQueue<>(),
-                        new ThreadConfiguration(getStaticThreadManager())
-                                .setThreadGroup(new ThreadGroup("virtual-cache-cleaners"))
-                                .setComponent("virtual-map")
-                                .setThreadName("cache-cleaner")
-                                .setExceptionHandler((t, ex) -> logger.error(
-                                        EXCEPTION.getMarker(), "Failed to purge unneeded key/mutationList pairs", ex))
-                                .buildFactory());
+        if (Boolean.getBoolean("syncCleaningPool")) {
+            cleaningPool = Runnable::run;
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:315]: This was needed for testing IIRC. Now that you allow to specify `Runnable::run` in a constructor you can remove the `if` and fix the test(s).

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 1 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:318]: Zero core pool size wouldn't give desired effect. This is why I used `allowCoreThreadTimeOut` instead. It's a safety net, really. Current implementation explicitly calls `shutdown` where it's needed. But even if we miss it in some cases, the thread will eventually die due to timeout.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 2 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:965]: Done

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 3 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:315]: Artem suggested passing null instead in cases where we don't need the cleaning pool at all. So, I implemented it.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 4]
```diff
@@ -817,6 +853,7 @@ private Stream<VirtualLeafBytes> dirtyLeaves(
      * 		if called on a cache that still allows dirty leaves to be added
      */
     public Stream<VirtualLeafBytes> deletedLeaves() {
+        assert cleaningPool != null : "deletedLeaves() called on a snapshot with no cleaning pool";
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java]: A similar assertion should be added to `dirtyLeaves()` and `dirtyHashesForFlush()` - both use the cleaning pool to filter mutations

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCacheTest.java [hunk 5]
```diff
@@ -2800,10 +2806,335 @@ void mergedCachesMemoryOverhead() {
...
+    void snapshotDeletedLeavesWorksWithSynchronousExecutor() {
+        // Set up a cache with some leaves, then delete one
+        cache.putLeaf(appleLeaf(1));
+        cache.putLeaf(bananaLeaf(2));
+        cache.putLeaf(cherryLeaf(3));
+
+        final VirtualNodeCache cache1 = cache.copy();
+        cache1.deleteLeaf(appleLeaf(1));
+
+        final VirtualNodeCache cache2 = cache1.copy();
```
Comment (thenswan) [inline platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCacheTest.java]: cache2 is not used

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 4 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java]: I'll add the assert to `dirtyLeaves` as it covers both cases.

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCacheTest.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCacheTest.java]: Fixed

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 4 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java]: `dirtyLeaves()` doesn't cover `dirtyHashesForFlush()`, it only covers `dirtyLeavesForFlush()` and `dirtyLeavesForHash()`

Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java]: In the latest implementation these asserts are no longer required. I removed them
